compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2494,11 +2494,19 @@ impl<'a> Parser<'a> {
         // Parse the arguments, starting out with `self` being allowed...
         let (mut params, _) = self.parse_paren_comma_seq(|p| {
             p.recover_diff_marker();
+            let meet_left_paren = p.check_noexpect(&token::OpenDelim(Delimiter::Parenthesis));
             let param = p.parse_param_general(req_name, first_param).or_else(|mut e| {
-                e.emit();
+                if p.capture_cfg {
```

Comment:
I don't understand why `p.capture_cfg` should determine whether we emit an error or not. Can you clarify this, please?

Reply:
Of course. This is because before changes, errors would be emitted repeatedly and duplicated errors are hided. But after this change, the close paren is be eaten if `p.capture_cfg` is true. This will emit different errors in the two phases, which may confuse users.

Reply:
> Could you remove this change from this PR? It is not necessary for fixing the ICE and `capture_cfg` shouldn't generally be used outside of token collection logic.

I did not use `capture_cfg` initially. Unfortunately, this affects some other existing diagnostics and leads to unexpected errors because it changes the behavior of parsing before token collection. For example, the output stderr for `tests/ui/parser/inverted-parameters.rs` will change without using `capture_cfg`:

test:
```rust
struct S;

impl S {
    fn foo(&self, &str bar) {}
    //~^ ERROR expected one of `:`, `@`
    //~| HELP declare the type after the parameter binding
    //~| SUGGESTION <identifier>: <type>
}

fn baz(S quux, xyzzy: i32) {}
//~^ ERROR expected one of `:`, `@`
//~| HELP declare the type after the parameter binding
//~| SUGGESTION <identifier>: <type>

fn one(i32 a b) {}
//~^ ERROR expected one of `:`, `@`

fn pattern((i32, i32) (a, b)) {}
//~^ ERROR expected one of `:`

fn fizz(i32) {}
//~^ ERROR expected one of `:`, `@`
//~| HELP if this is a parameter name, give it a type
//~| HELP if this is a `self` type, give it a parameter name
//~| HELP if this is a type, explicitly ignore the parameter name

fn missing_colon(quux S) {}
//~^ ERROR expected one of `:`, `@`
//~| HELP declare the type after the parameter binding
//~| SUGGESTION <identifier>: <type>

fn main() {}
```

stderr (before):
```
error: expected one of `:`, `@`, or `|`, found `bar`
  --> $DIR/inverted-parameters.rs:4:24
   |
LL |     fn foo(&self, &str bar) {}
   |                   -----^^^
   |                   |    |
   |                   |    expected one of `:`, `@`, or `|`
   |                   help: declare the type after the parameter binding: `<identifier>: <type>`

error: expected one of `:`, `@`, or `|`, found `quux`
  --> $DIR/inverted-parameters.rs:10:10
   |
LL | fn baz(S quux, xyzzy: i32) {}
   |        --^^^^
   |        | |
   |        | expected one of `:`, `@`, or `|`
   |        help: declare the type after the parameter binding: `<identifier>: <type>`

error: expected one of `:`, `@`, or `|`, found `a`
  --> $DIR/inverted-parameters.rs:15:12
   |
LL | fn one(i32 a b) {}
   |            ^ expected one of `:`, `@`, or `|`

error: expected one of `:` or `|`, found `(`
  --> $DIR/inverted-parameters.rs:18:23
   |
LL | fn pattern((i32, i32) (a, b)) {}
   |                       ^ expected one of `:` or `|`

error: expected one of `:`, `@`, or `|`, found `)`
  --> $DIR/inverted-parameters.rs:21:12
   |
LL | fn fizz(i32) {}
   |            ^ expected one of `:`, `@`, or `|`
   |
   = note: anonymous parameters are removed in the 2018 edition (see RFC 1685)
help: if this is a `self` type, give it a parameter name
   |
LL | fn fizz(self: i32) {}
   |         +++++
help: if this is a parameter name, give it a type
   |
LL | fn fizz(i32: TypeName) {}
   |            ++++++++++
help: if this is a type, explicitly ignore the parameter name
   |
LL | fn fizz(_: i32) {}
   |         ++

error: expected one of `:`, `@`, or `|`, found `S`
  --> $DIR/inverted-parameters.rs:27:23
   |
LL | fn missing_colon(quux S) {}
   |                  -----^
   |                  |    |
   |                  |    expected one of `:`, `@`, or `|`
   |                  help: declare the type after the parameter binding: `<identifier>: <type>`

error: aborting due to 6 previous errors
```

stderr (after):
```
error: expected one of `:`, `@`, or `|`, found `bar`
  --> $DIR/inverted-parameters.rs:4:24
   |
LL |     fn foo(&self, &str bar) {}
   |                   -----^^^
   |                   |    |
   |                   |    expected one of `:`, `@`, or `|`
   |                   help: declare the type after the parameter binding: `<identifier>: <type>`

error: expected one of `:`, `@`, or `|`, found `quux`
  --> $DIR/inverted-parameters.rs:10:10
   |
LL | fn baz(S quux, xyzzy: i32) {}
   |        --^^^^
   |        | |
   |        | expected one of `:`, `@`, or `|`
   |        help: declare the type after the parameter binding: `<identifier>: <type>`

error: expected one of `:`, `@`, or `|`, found `a`
  --> $DIR/inverted-parameters.rs:15:12
   |
LL | fn one(i32 a b) {}
   |            ^ expected one of `:`, `@`, or `|`

error: expected one of `:` or `|`, found `(`
  --> $DIR/inverted-parameters.rs:18:23
   |
LL | fn pattern((i32, i32) (a, b)) {}
   |                       ^ expected one of `:` or `|`

error: expected parameter name, found `{`
  --> $DIR/inverted-parameters.rs:18:31
   |
LL | fn pattern((i32, i32) (a, b)) {}
   |                               ^ expected parameter name

error: expected one of `)` or `,`, found `{`
  --> $DIR/inverted-parameters.rs:18:31
   |
LL | fn pattern((i32, i32) (a, b)) {}
   |                              -^ expected one of `)` or `,`
   |                              |
   |                              help: missing `,`

error: expected one of `)`, `,`, `->`, `where`, or `{`, found `<eof>`
  --> $DIR/inverted-parameters.rs:32:12
   |
LL | fn main() {}
   |            ^ expected one of `)`, `,`, `->`, `where`, or `{`

error: aborting due to 7 previous errors
```

Reply:
Could you make this change and commit the result?
So I could look at a proper diff instead of a comment.

Reply:
Sure :)

Reply:
The new commit is pushed